pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -176,7 +175,15 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 		// and may not have had all of the assumed default specified.
 		pr.SetDefaults(contexts.WithUpgradeViaDefaulting(ctx))
 
-		c.updatePipelineResults(ctx, pr, getPipelineFunc)
+		if _, pipelineSpec, err := resources.GetPipelineData(ctx, pr, getPipelineFunc); err != nil {
+			logger.Errorf("Failed to get Pipeline Spec to process Pipeline Results: %v", err)
+			pr.Status.MarkFailed(ReasonCouldntGetPipeline,
```

Comment:
Question for @pritidesai or @afrittoli - given that we are in a block guarded by `pr.IsDone()`, is it too late to call `pr.Status.MarkFailed()`?  I hadn't noticed this prior to making these changes but the old code did this too (see `updatePipelineResults` in the diff.

Reply:
Good point, `pr.IsDone` declares pipeline either `success` or `failure`.  `markFailed` can switch the `success` to `failure` with setting its `completionTime`. This can definitely cause issue since the `pipeline` would be marked as finished with the `completionTime` but it's still running.

Its missing call to `finishReconcileUpdateEmitEvents` which will make sure the `pipeline` actually finishes and doesnt continue execution after `MarkFailed`.

```
		pr.Status.MarkFailed(ReasonCouldntGetPipeline, "Failed to get Pipeline Spec to process Pipeline Results %s/%s: %s", pr.Namespace, pr.Name, err)
		return c.finishReconcileUpdateEmitEvents(ctx, pr, before, nil)
```

Reply:
this brings us back to similar discussion with task results, all the tasks executed successfully but a particular pipeline result which was being referenced doesn't exist so declare pipeline failed 🤔

Reply:
I'm going to remove the call to `MarkFailed` here I think.

Reply:
I've removed the call to `MarkFailed` and now instead am using the controller's Recorder (if available) to record a warning event (in addition to logging an error from the controller).

Reply:
FYI, as part of the Run integration with pipelines I tried to have the e2e test verify the pipeline results.  I had to take that out because there was a race condition between the test case waking up (seeing the PipelineRun was done) and the controller updating the PipelineRun with the results.  I think the results should be added to the PR at the same time it is marked done.  I'm just bringing this up for your awareness; it's up to you whether you want to tackle it in this PR.

Reply:
that's a good point @GregDritschler 

yup, we dont have to tackle everything in this PR, sorry @sbwsg for polluting this PR.

If I am not mistaken, taskRun has the similar behavior where a taskRun is marked as successful along with its completion time before populating task results:

https://github.com/tektoncd/pipeline/blob/05d67e427c722a2a57e58328d7097e21429b7524/pkg/pod/status.go#L272-L282

https://github.com/tektoncd/pipeline/blob/05d67e427c722a2a57e58328d7097e21429b7524/pkg/pod/status.go#L109-L113

After a taskRun is marked as successful, taskRun status is updated with task results (through [setTaskRunStatusBasedOnStepStatus](https://github.com/tektoncd/pipeline/blob/05d67e427c722a2a57e58328d7097e21429b7524/pkg/pod/status.go#L132)):

https://github.com/tektoncd/pipeline/blob/05d67e427c722a2a57e58328d7097e21429b7524/pkg/pod/status.go#L162-L164

may be worth a discussion in API WG? 😜

Reply:
Thanks for the feedback, I'm going to back out the change and re-introduce `MarkFailed` since it's the existing behaviour and we can discuss further.

Reply:
Done - MarkFailed is re-introduced.